imports/composite_redolog_manager.go

```diff
@@ -0,0 +1,330 @@
+
+			err = s.kafkaReader.ConsumeFrom(offsetFrom)
+			if err != nil {
+				utils.GetLogger().Fatal(err)
+			}
+
+			fn := s.kafkaReader.NextUpsertBatch()
+			for {
+				batch, kafkaOffset := fn()
+				if batch == nil {
+					break
+				}
+
+				if !s.recoveryDone && kafkaOffset >= offsetTo {
+					s.setRecoveryDone()
+					utils.GetLogger().With("action", "ingestion", "table", s.Table, "shard", s.Shard,
+						"offset", offsetTo, "batchRecovered", s.msgRecovered).Info("Recovery from kafka done, start normal ingestion")
+				}
+				if s.recoveryDone {
+					err = s.applyUpsertBatchWithLogging(batch, kafkaOffset)
+				} else {
+					redoLogFile, batchOffset := s.kafkaReader.getFileOffset(kafkaOffset)
+					err = s.applyUpsertBatchOnly(batch, redoLogFile, batchOffset)
+				}
+
+				if err != nil {
+					s.msgError++
+					utils.GetLogger().With("action", "ingestion", "table", s.Table, "shard", s.Shard,
+						"offset", kafkaOffset).Error("Error save upsertbatch to store")
+					utils.GetReporter(s.Table, s.Shard).GetCounter(utils.IngestedErrorBatches).Inc(1)
+				}
+			}
+			utils.GetLogger().With("action", "ingestion", "table", s.Table, "shard", s.Shard).Warn("Kafka ingestion stopped")
+		}
+		s.done = true
+	}()
+
+	return nil
+}
+
+func (s *CompositeRedologManager) setRecoveryDone() {
+	// no locking here as this will always be sequential read/write
+	s.recoveryDone = true
+	s.readyChan <- true
+}
+
+// applyUpsertBatchOnly is used for replay redolog, so no redolog write necessary
+func (s *CompositeRedologManager) applyUpsertBatchOnly(batch *common.UpsertBatch, redoLogFile int64, batchOffset uint32) error {
+
+	utils.GetReporter(s.Table, s.Shard).GetCounter(utils.IngestedUpsertBatches).Inc(1)
+	utils.GetReporter(s.Table, s.Shard).GetGauge(utils.UpsertBatchSize).Update(float64(len(batch.GetBuffer())))
+
+	s.Lock()
+	defer s.Unlock()
+
+	s.msgRecovered++
+	skipBackfillRows := redoLogFile < s.redoFilePersisited ||
+		(redoLogFile == s.redoFilePersisited && batchOffset <= s.offsetPersisted)
+
+	s.UpdateMaxEventTime(0, redoLogFile)
+	if s.storeFunc != nil {
+		return s.storeFunc(batch, redoLogFile, batchOffset, skipBackfillRows)
+	}
+	return nil
+}
+
+// applyUpsertBatchWithLogging is used for normal ingestion (kafka or http)
+func (s *CompositeRedologManager) applyUpsertBatchWithLogging(batch *common.UpsertBatch, offsetInSource int64) error {
+
+	utils.GetReporter(s.Table, s.Shard).GetCounter(utils.IngestedUpsertBatches).Inc(1)
+	utils.GetReporter(s.Table, s.Shard).GetGauge(utils.UpsertBatchSize).Update(float64(len(batch.GetBuffer())))
+
+	s.Lock()
+	defer s.Unlock()
+
+	s.msgReceived++
+	redoLogFile, batchOffset := s.AppendToRedoLog(batch, offsetInSource)
+	if s.storeFunc != nil {
+		return s.storeFunc(batch, redoLogFile, batchOffset, false)
+	}
+	return nil
+}
+
+// WaitForRecoveryDone block call to wait for recovery finish
+func (s *CompositeRedologManager) WaitForRecoveryDone() {
+	select {
+	case <-s.readyChan:
```

Comment:
no need to select for a single channel, or I would rather using a conditional variable